src/olympia/shelves/serializers.py

```diff
@@ -6,21 +6,66 @@
 
 from olympia import amo
 from olympia.addons.views import AddonSearchView
+from olympia.amo.reverse import reverse
+from olympia.amo.templatetags.jinja_helpers import absolutify
 from olympia.api.fields import (
     AbsoluteOutgoingURLField,
     GetTextTranslationSerializerField,
     ReverseChoiceField,
 )
+from olympia.api.utils import is_gate_active
 from olympia.bandwagon.views import CollectionAddonViewSet
-from olympia.hero.serializers import CTAField
 
 from .models import Shelf
 
 
-class ShelfFooterField(CTAField):
+class ShelfFooterField(serializers.Serializer):
     url = AbsoluteOutgoingURLField(source='footer_pathname')
     text = GetTextTranslationSerializerField(source='footer_text')
 
+    def to_representation(self, obj):
+        data = super().to_representation(obj)
+        request = self.context.get('request', None)
+        # when 'wrap-outgoing-parameter' is on url is a flat string already
+        is_flat_url = request and is_gate_active(request, 'wrap-outgoing-parameter')
+
+        url = data.get('url')
+        if not url and obj.endpoint in (
+            Shelf.Endpoints.SEARCH,
+            Shelf.Endpoints.RANDOM_TAG,
+        ):
+            search_url = reverse('search.search')
+            query_string = '&'.join(
+                f'{key}={value}' for key, value in obj.get_param_dict().items()
+            )
+            fallback = absolutify(f'{search_url}?{query_string}')
+            url = (
+                {'url': fallback, 'outgoing': fallback} if not is_flat_url else fallback
+            )
+        elif not url and obj.endpoint == Shelf.Endpoints.COLLECTIONS:
+            fallback = absolutify(
+                reverse(
+                    'collections.detail',
+                    kwargs={
+                        'user_id': str(settings.TASK_USER_ID),
+                        'slug': obj.criteria,
+                    },
+                )
+            )
+            url = (
+                {'url': fallback, 'outgoing': fallback} if not is_flat_url else fallback
+            )
+        # text = data.get('text')
+
+        if is_flat_url:
+            return url
+        else:
+            return {
+                **data,
```

Comment:
What would this `**data` be? If we don't need to worry about it when `is_flat_url` is `True`, why do we need to worry about it here? I guess it's just being safe in case there is something there?

Reply:
`data` will contain `text` too (and other fields if they're added to the serializer later) but I see now I forgot to do the same when `flat_url` is true above!

Reply:
fixed in https://github.com/mozilla/addons-server/pull/17550/commits/332ed63a8aa3d52f06a8fa8db6fc9ead32858507 - and changed the tests to better check that the output is what we expect (before `ShelfSerializer` was using `CTAField` so we partly relied on those tests to cover all the edge cases with the more unusual api gates)